File: clients/windows/src/Vertex.Core/Transport/Transport.cs
namespace Vertex.Core.Transport;

/// <summary>
/// Coarse transport state for UI reporting. Mirrors Swift
/// <c>TransportState</c> and Kotlin <c>TransportState</c>.
/// </summary>
public abstract record TransportState
{
    public sealed record Disconnected : TransportState;
    public sealed record Connecting(string Broker) : TransportState;
    public sealed record Connected(string Broker)  : TransportState;
    public sealed record Reconnecting(string Broker, int Attempt) : TransportState;
}

/// <summary>Errors raised by <see cref="IMqttTransport"/> public methods.</summary>
public sealed class TransportException : Exception
{
    public TransportException(string message) : base(message) { }

    public static TransportException ConnectTimeout() => new("Transport connect timeout reached.");
    public static TransportException NoBrokers()       => new("No brokers configured.");
}

/// <summary>
/// Abstract pub/sub message transport for the Vertex data plane. Today the
/// only implementation is <see cref="MqttTransport"/>; the contract is
/// minimal so a future NATS / gRPC / QUIC backend can drop in without
/// touching the data-plane.
/// </summary>
public interface IMqttTransport : IAsyncDisposable
{
    /// <summary>True once we received a successful CONNACK on the current broker.</summary>
    bool IsReady { get; }

    /// <summary>Hostname of the currently active broker, or <c>null</c> when disconnected.</summary>
    string? CurrentBroker { get; }

    /// <summary>Async stream of state transitions for the host UI / logger.</summary>
    IAsyncEnumerable<TransportState> StateUpdates { get; }

    /// <summary>Begin connecting; returns when the first CONNACK arrives or the
    /// connect deadline expires (then throws <see cref="TransportException"/>).</summary>
    Task StartAsync(CancellationToken ct = default);

    /// <summary>Graceful disconnect. Cancels reconnection. Idempotent.</summary>
    Task StopAsync();

    /// <summary>Publish QoS 0. Drops silently if not connected.</summary>
    void Publish(string topic, ReadOnlyMemory<byte> payload);

    /// <summary>Register a handler for messages whose topic matches <paramref name="pattern"/>.
    /// Subscriptions persist across reconnects.</summary>
    void Subscribe(string pattern, Action<string, byte[]> handler);

    /// <summary>Stop dispatching messages for <paramref name="pattern"/> locally.
    /// Does not currently send MQTT UNSUBSCRIBE — broker keeps delivering, dispatch becomes no-op.</summary>
    void Unsubscribe(string pattern);

    /// <summary>Tear down the current connection and reconnect immediately, bypassing backoff.</summary>
    void ForceReconnect(string reason);

    /// <summary>Force a fresh PINGREQ ahead of cadence — used on system wake / NetworkChange hints.</summary>
    void CheckLiveness(string reason);

    /// <summary>Wait until <see cref="IsReady"/> or throw <see cref="TransportException.ConnectTimeout"/>.</summary>
    Task WaitReadyAsync(TimeSpan timeout, CancellationToken ct = default);
}

File: clients/shared/VertexCore/Sources/VertexCore/Transport/Transport.swift
import Foundation

/// Abstract message transport for the Vertex data plane.
///
/// Concrete implementations live alongside this file (today: `MQTTTransport`;
/// future candidates: NATS, gRPC bidi stream, QUIC datagram). The protocol
/// is intentionally minimal — universally-supported pub/sub primitives only.
/// Capabilities that vary by transport (retained messages, last-will,
/// schema validation) sit in separate protocols (e.g. `Retainer`) so that
/// consumers needing a capability assert it explicitly:
///
/// ```swift
/// if let r = transport as? Retainer {
///     r.publishRetained(topic: ..., payload: ...)
/// }
/// ```
///
/// On Apple platforms we use class-bound protocol existentials so the
/// runtime cost stays low (no boxing for value-type implementations because
/// there are none in scope).
public protocol Transport: AnyObject, Sendable {
    /// Start connecting. Returns when the first CONNACK arrives, or throws
    /// on timeout / authentication failure.
    func start() async throws

    /// Graceful disconnect. Cancels any pending reconnection.
    func stop()

    /// Publish a message to the given topic. Fire-and-forget — drops
    /// silently when not connected (caller is expected to retry via the
    /// application protocol if reliability is needed; the data plane uses
    /// QoS 0 semantics).
    func publish(topic: String, payload: Data)

    /// Register a handler for messages whose topic matches the pattern.
    /// Pattern semantics are transport-defined (MQTT wildcards `+`/`#` for
    /// the MQTT impl). Subscriptions are persisted across reconnects.
    func subscribe(pattern: String, handler: @escaping @Sendable (String, Data) -> Void)

    /// Wait until the transport reaches the connected state, or `timeout`
    /// elapses (then throws `TransportError.connectTimeout`).
    func waitReady(timeout: Duration) async throws

    /// Tear down the underlying connection and reconnect immediately,
    /// bypassing the backoff schedule. Implementations should debounce
    /// rapid successive calls. Currently used internally by MQTT
    /// implementations when application-level keepalive (PINGRESP
    /// timeout) detects a dead link.
    func forceReconnect(reason: String)

    /// True when the transport is connected and operational.
    var isReady: Bool { get }

    /// Hostname of the broker we're currently connected to, or nil.
    var currentBroker: String? { get }

    /// Async stream of state transitions, for UI reporting.
    var stateUpdates: AsyncStream<TransportState> { get }
}

/// Optional capability for transports that support state-style messages
/// where new subscribers receive the last value published on the topic.
/// MQTT supports it natively (retained); NATS Core does not (use JetStream
/// KV); WebRTC has no equivalent. No Swift consumer publishes retained
/// today — this protocol exists for symmetry with the Go side and future
/// in-process exit implementations.
public protocol Retainer: AnyObject, Sendable {
    func publishRetained(topic: String, payload: Data)
}

/// Connection state for UI reporting. Mirrors the Go-side states.
public enum TransportState: Sendable {
    case disconnected
    case connecting(broker: String)
    case connected(broker: String)
    case reconnecting(broker: String, attempt: Int)
}

public enum TransportError: Error, Sendable {
    case connectTimeout
    case noBrokers
}
